File: mParticle-iterable/MPKitIterable.m
#import "MPKitIterable.h"
#import <objc/runtime.h>
#import <objc/message.h>
@import IterableSDK;


@interface MPKitIterable() <IterableURLDelegate>
@end

@implementation MPKitIterable

@synthesize kitApi = _kitApi;
static IterableConfig *_customConfig = nil;
static NSURL *_clickedURL = nil;
static BOOL *prefersUserId = FALSE;

+ (BOOL *)prefersUserId {
    return prefersUserId;
}

+ (void)setPrefersUserId:(BOOL)prefers {
    prefersUserId = &prefers;
}

+ (NSNumber *)kitCode {
    return @1003;
}

+ (void)load {
    MPKitRegister *kitRegister = [[MPKitRegister alloc] initWithName:@"Iterable" className:@"MPKitIterable"];
    [MParticle registerExtension:kitRegister];
}

+ (void)setCustomConfig:(IterableConfig *_Nullable)config {
    _customConfig = config;
}

- (BOOL)handleIterableURL:(NSURL *)url context:(IterableActionContext *)context {
    BOOL result = YES;
    if (_customConfig.urlDelegate != nil && [((NSObject *)_customConfig.urlDelegate) respondsToSelector:@selector(handleIterableURL:context:)]) {
        result = [_customConfig.urlDelegate handleIterableURL:url context: context];
    } else if (_customConfig.urlDelegate != nil) {
        NSLog(@"mParticle -> Error: Iterable urlDelegate was set in custom config but didn't respond to the selector 'handleIterableURL:context:'");
    }

    NSString *destinationURL = url.absoluteString;
    NSDictionary *getAndTrackParams = nil;
    NSString *clickedUrlString = [NSString stringWithString:_clickedURL.absoluteString];
    _clickedURL = nil;
    if (!destinationURL || [clickedUrlString isEqualToString:destinationURL]) {
        getAndTrackParams = [[NSDictionary alloc] initWithObjectsAndKeys: clickedUrlString, IterableClickedURLKey, nil];
    } else {
        getAndTrackParams = [[NSDictionary alloc] initWithObjectsAndKeys: destinationURL, IterableDestinationURLKey, clickedUrlString, IterableClickedURLKey, nil];
    }
    
    MPAttributionResult *attributionResult = [[MPAttributionResult alloc] init];
    attributionResult.linkInfo = getAndTrackParams;

    [self->_kitApi onAttributionCompleteWithResult:attributionResult error:nil];
    return true;
}

#pragma mark - MPKitInstanceProtocol methods

#pragma mark Kit instance and lifecycle
- (MPKitExecStatus *)didFinishLaunchingWithConfiguration:(NSDictionary *)configuration {
    MPKitExecStatus *execStatus = nil;
    
    _configuration = configuration;
    
    [self start];
    
    execStatus = [[MPKitExecStatus alloc] initWithSDKCode:[[self class] kitCode] returnCode:MPKitReturnCodeSuccess];
    return execStatus;
}

- (void)start {
    static dispatch_once_t kitPredicate;
    
    dispatch_once(&kitPredicate, ^{
        
        NSString *apiKey = self.configuration[@"apiKey"];
        NSString *apnsProdIntegrationName = self.configuration[@"apnsProdIntegrationName"];
        NSString *apnsSandboxIntegrationName = self.configuration[@"apnsSandboxIntegrationName"];
        NSString *userIdField = self.configuration[@"userIdField"];
        self.mpidEnabled = [userIdField isEqualToString:@"mpid"];

        IterableConfig *config = _customConfig;
        if (!config) {
            config = [[IterableConfig alloc] init];
        }
        config.pushIntegrationName = apnsProdIntegrationName;
        config.sandboxPushIntegrationName = apnsSandboxIntegrationName;
        config.urlDelegate = self;
        
        [IterableAPI initializeWithApiKey:apiKey config:config];
        [self initIntegrationAttributes];

        self->_started = YES;
        
        dispatch_async(dispatch_get_main_queue(), ^{
            NSDictionary *userInfo = @{mParticleKitInstanceKey:[[self class] kitCode]};
            
            [[NSNotificationCenter defaultCenter] postNotificationName:mParticleKitDidBecomeActiveNotification
                                                                object:nil
                                                              userInfo:userInfo];
        });
    });
}

#pragma mark Application
- (nonnull MPKitExecStatus *)continueUserActivity:(nonnull NSUserActivity *)userActivity restorationHandler:(void(^ _Nonnull)(NSArray * _Nullable restorableObjects))restorationHandler {
    _clickedURL = userActivity.webpageURL;

    if (_clickedURL != nil) {
        [IterableAPI handleUniversalLink:_clickedURL];
    }

    MPKitExecStatus *execStatus = [[MPKitExecStatus alloc] initWithSDKCode:[MPKitIterable kitCode] returnCode:MPKitReturnCodeSuccess];
    return execStatus;
}

- (void)initIntegrationAttributes {
    NSDictionary *integrationAttributes = @{
            @"Iterable.sdkVersion": IterableAPI.sdkVersion
    };
    [[MParticle sharedInstance] setIntegrationAttributes:integrationAttributes forKit:MPKitIterable.kitCode];
}

- (NSString *)getUserEmail:(FilteredMParticleUser *)user {
    return user.userIdentities[@(MPUserIdentityEmail)];
}

- (NSString *)getCustomerId:(FilteredMParticleUser *)user {
    return user.userIdentities[@(MPUserIdentityCustomerId)];
}

- (MPKitExecStatus *)onLoginComplete:(FilteredMParticleUser *)user request:(FilteredMPIdentityApiRequest *)request {
    [self updateIdentity:user];
    return [[MPKitExecStatus alloc] initWithSDKCode:[[self class] kitCode] returnCode:MPKitReturnCodeSuccess];
}

- (nonnull MPKitExecStatus *)onLogoutComplete:(nonnull FilteredMParticleUser *)user request:(nonnull FilteredMPIdentityApiRequest *)request {
    [self updateIdentity:user];
    return [[MPKitExecStatus alloc] initWithSDKCode:[[self class] kitCode] returnCode:MPKitReturnCodeSuccess];
}

- (MPKitExecStatus *)onIdentifyComplete:(FilteredMParticleUser *)user request:(FilteredMPIdentityApiRequest *)request {
    [self updateIdentity:user];
    return [[MPKitExecStatus alloc] initWithSDKCode:[[self class] kitCode] returnCode:MPKitReturnCodeSuccess];
}

- (NSString *)getPlaceholderEmail:(NSString *)userId {
    if (userId.length > 0) {
        return [NSString stringWithFormat:@"%@@placeholder.email", userId];
    } else {
        return nil;
    }
}

- (NSString *)getUserId:(FilteredMParticleUser *)user {
    NSString *id = nil;
    if (self.mpidEnabled) {
        if (user.userId.longValue != 0) {
            id = user.userId.stringValue;
        }
    } else {
        id = [[[[UIDevice currentDevice] identifierForVendor] UUIDString] lowercaseString];

        if (!id.length) {
            id = [[self advertiserId] lowercaseString];
        }

        if (!id.length) {
            id = [self getCustomerId:user];
        }

        if (!id.length) {
            id = [[[MParticle sharedInstance] identity] deviceApplicationStamp];
        }
    }
    return id;
}

- (void)updateIdentity:(FilteredMParticleUser *)user {
    NSString *userId = [self getUserId:user];
    
    if (prefersUserId) {
        [IterableAPI setUserId:userId];
        return;
    }
    
    NSString *email = [self getUserEmail:user];
    NSString *placeholderEmail = [self getPlaceholderEmail:userId];
    
    if (email != nil && email.length > 0) {
        [IterableAPI setEmail:email];
    } else if (placeholderEmail != nil && placeholderEmail.length > 0) {
        [IterableAPI setEmail:placeholderEmail];
    } else {
        // No identifier, log out
        [IterableAPI setEmail:nil];
    }
}

- (FilteredMParticleUser *)currentUser {
    return [[self kitApi] getCurrentUserWithKit:self];
}

- (MPKitExecStatus *)setDeviceToken:(NSData *)deviceToken {
    [IterableAPI registerToken:deviceToken];
    return [[MPKitExecStatus alloc] initWithSDKCode:[[self class] kitCode] returnCode:MPKitReturnCodeSuccess];
}

#if TARGET_OS_IOS == 1 && __IPHONE_OS_VERSION_MAX_ALLOWED >= __IPHONE_10_0
- (MPKitExecStatus *)userNotificationCenter:(UNUserNotificationCenter *)center didReceiveNotificationResponse:(UNNotificationResponse *)response API_AVAILABLE(ios(10.0)) {
    [IterableAppIntegration userNotificationCenter:center didReceiveNotificationResponse:response withCompletionHandler:^{}];
    return [[MPKitExecStatus alloc] initWithSDKCode:[[self class] kitCode] returnCode:MPKitReturnCodeSuccess];
}
#endif
 
- (MPKitExecStatus *)receivedUserNotification:(NSDictionary *)userInfo {
    [IterableAppIntegration application:UIApplication.sharedApplication didReceiveRemoteNotification:userInfo fetchCompletionHandler:^(UIBackgroundFetchResult result){}];
    return [[MPKitExecStatus alloc] initWithSDKCode:[[self class] kitCode] returnCode:MPKitReturnCodeSuccess];
}

#pragma mark Accessors
- (NSString *)advertiserId {
    NSString *advertiserId = nil;
    Class MPIdentifierManager = NSClassFromString(@"ASIdentifierManager");
    if (MPIdentifierManager) {
#pragma clang diagnostic push
#pragma clang diagnostic ignored "-Wundeclared-selector"
#pragma clang diagnostic push
#pragma clang diagnostic ignored "-Warc-performSelector-leaks"
        SEL selector = NSSelectorFromString(@"sharedManager");
        id<NSObject> adIdentityManager = [MPIdentifierManager performSelector:selector];

        selector = NSSelectorFromString(@"isAdvertisingTrackingEnabled");
        BOOL advertisingTrackingEnabled = ((BOOL (*)(id, SEL))objc_msgSend)(adIdentityManager, selector);
        if (advertisingTrackingEnabled) {
            selector = NSSelectorFromString(@"advertisingIdentifier");
            advertiserId = [[adIdentityManager performSelector:selector] UUIDString];
        }
#pragma clang diagnostic pop
#pragma clang diagnostic pop
    }

    return advertiserId;
}

@end

File: mParticle-Iterable/MPKitIterable.swift
import Foundation
import mParticle_Apple_SDK
import IterableSDK

@objcMembers public class MPKitIterable: NSObject, MPKitProtocol, IterableURLDelegate {
    
    private var _kitApi: MPKitAPI?
    public var configuration: [AnyHashable: Any] = [:]
    public var launchOptions: [AnyHashable: Any]?
    public var started: Bool = false
    public var userAttributes: [String: Any]?
    public var userIdentities: [[String: Any]]?
    public var mpidEnabled: Bool = false
    
    static private var customConfig: IterableConfig?
    static private var customUrlDelegate: IterableURLDelegate?
    static private var clickedURL: URL?
    static private var _prefersUserId: Bool = false
    
    public static func setCustomConfig(_ config: IterableConfig?) {
        customConfig = config
        customUrlDelegate = config?.urlDelegate
    }
    
    public static func prefersUserId() -> Bool {
        return _prefersUserId
    }
    
    public static func kitCode() -> NSNumber {
        return 1003
    }
    
    public static func setPrefersUserId(_ prefers: Bool) {
        _prefersUserId = prefers
    }
    
    public func handle(iterableURL url: URL, inContext context: IterableSDK.IterableActionContext) -> Bool {
        var result = true
        
        if MPKitIterable.customUrlDelegate === self {
            print("mParticle -> Error: Iterable urlDelegate was set in custom config but points to the MPKitIterable instance. It should be a different object.")
        } else if let customUrlDelegate = MPKitIterable.customUrlDelegate as? NSObject, customUrlDelegate.responds(to: #selector(IterableURLDelegate.handle(iterableURL:inContext:))) {
            if let customResult = customUrlDelegate.perform(#selector(IterableURLDelegate.handle(iterableURL:inContext:)), with: url, with: context)?.takeUnretainedValue() as? Bool {
                result = customResult
            }
        } else if MPKitIterable.customUrlDelegate != nil {
            print("mParticle -> Error: Iterable urlDelegate was set in custom config but didn't respond to the selector 'handleIterableURL:context:'")
        }
        
        let destinationURL = url.absoluteString
        var getAndTrackParams: [String: Any]?
        let clickedUrlString = MPKitIterable.clickedURL?.absoluteString ?? ""
        MPKitIterable.clickedURL = nil
        
        if destinationURL.isEmpty || clickedUrlString == destinationURL {
            getAndTrackParams = [IterableClickedURLKey: clickedUrlString]
        } else {
            getAndTrackParams = [IterableDestinationURLKey: destinationURL, IterableClickedURLKey: clickedUrlString]
        }
        
        let attributionResult = MPAttributionResult()
        attributionResult.linkInfo = getAndTrackParams ?? [:]
        
        _ = _kitApi?.onAttributionComplete(with: attributionResult, error: nil)
        
        return result
    }
    
    // MARK: - MPKitInstanceProtocol methods
    
    // MARK: Kit instance and lifecycle
    public func didFinishLaunching(withConfiguration configuration: [AnyHashable: Any]) -> MPKitExecStatus {
        self.configuration = configuration
        start()
        return MPKitExecStatus(sdkCode: MPKitIterable.kitCode(), returnCode: .success)
    }
    
    public func start() {
        struct OnceToken {
            static var token = false
        }
        
        if !OnceToken.token {
            if let apiKey = self.configuration["apiKey"] as? String,
               let apnsProdIntegrationName = self.configuration["apnsProdIntegrationName"] as? String,
               let apnsSandboxIntegrationName = self.configuration["apnsSandboxIntegrationName"] as? String,
               let userIdField = self.configuration["userIdField"] as? String {
                
                self.mpidEnabled = userIdField == "mpid"
                
                var config = Self.customConfig
                if config == nil {
                    config = IterableConfig()
                }
                
                config?.pushIntegrationName = apnsProdIntegrationName
                config?.sandboxPushIntegrationName = apnsSandboxIntegrationName
                config?.urlDelegate = self
                
                if let config = config {
                    IterableAPI.initialize(apiKey: apiKey, config: config)
                    initIntegrationAttributes()
                    
                    self.started = true
                    
                    DispatchQueue.main.async {
                        let userInfo = [mParticleKitInstanceKey: MPKitIterable.kitCode()]
                        
                        NotificationCenter.default.post(name: NSNotification.Name(NSNotification.Name.mParticleKitDidBecomeActive.rawValue),
                                                        object: nil,
                                                        userInfo: userInfo)
                    }
                }
            }
        }
        
        func continueUserActivity(_ userActivity: NSUserActivity, restorationHandler: @escaping ([Any]?) -> Void) -> MPKitExecStatus {
            MPKitIterable.clickedURL = userActivity.webpageURL
            
            if let clickedURL = MPKitIterable.clickedURL {
                IterableAPI.handle(universalLink: clickedURL)
            }
            
            return MPKitExecStatus(sdkCode: MPKitIterable.kitCode(), returnCode: .success)
        }
        
        func initIntegrationAttributes() {
            let integrationAttributes = [
                "Iterable.sdkVersion": IterableAPI.sdkVersion
            ]
            MParticle.sharedInstance().setIntegrationAttributes(integrationAttributes, forKit: MPKitIterable.kitCode())
        }
        
        func getUserEmail(_ user: FilteredMParticleUser) -> String? {
            return user.userIdentities[NSNumber(value: MPUserIdentity.email.rawValue)]
        }
        
        func getCustomerId(_ user: FilteredMParticleUser) -> String? {
            return user.userIdentities[NSNumber(value: MPUserIdentity.customerId.rawValue)]
        }
        
        func onLoginComplete(_ user: FilteredMParticleUser, request: FilteredMPIdentityApiRequest) -> MPKitExecStatus {
            updateIdentity(user)
            return MPKitExecStatus(sdkCode: MPKitIterable.kitCode(), returnCode: .success)
        }
        
        func onLogoutComplete(_ user: FilteredMParticleUser, request: FilteredMPIdentityApiRequest) -> MPKitExecStatus {
            updateIdentity(user)
            return MPKitExecStatus(sdkCode: MPKitIterable.kitCode(), returnCode: .success)
        }
        
        func onIdentifyComplete(_ user: FilteredMParticleUser, request: FilteredMPIdentityApiRequest) -> MPKitExecStatus {
            updateIdentity(user)
            return MPKitExecStatus(sdkCode: MPKitIterable.kitCode(), returnCode: .success)
        }
        
        func getUserId(_ user: FilteredMParticleUser) -> String? {
            var userId: String? = nil
            if mpidEnabled {
                if user.userId != 0 {
                    userId = user.userId.stringValue
                }
            } else {
                userId = UIDevice.current.identifierForVendor?.uuidString.lowercased()
                
                if userId?.isEmpty ?? true {
                    userId = advertiserId()?.lowercased()
                }
                
                if userId?.isEmpty ?? true {
                    userId = getCustomerId(user)
                }
                
                if userId?.isEmpty ?? true {
                    userId = MParticle.sharedInstance().identity.deviceApplicationStamp
                }
            }
            return userId
        }
        
        func getPlaceholderEmail(_ userId: String?) -> String? {
            guard let userId = userId, !userId.isEmpty else {
                return nil
            }
            return "\(userId)@placeholder.email"
        }
        
        func updateIdentity(_ user: FilteredMParticleUser) {
            if let userId = getUserId(user), MPKitIterable._prefersUserId {
                IterableAPI.setUserId(userId)
                return
            }
            
            let email = getUserEmail(user)
            let placeholderEmail = getPlaceholderEmail(getUserId(user))
            
            if let email = email, !email.isEmpty {
                IterableAPI.setEmail(email)
            } else if let placeholderEmail = placeholderEmail, !placeholderEmail.isEmpty {
                IterableAPI.setEmail(placeholderEmail)
            } else {
                // No identifier, log out
                IterableAPI.setEmail(nil)
            }
        }
        
        func currentUser() -> FilteredMParticleUser? {
            return _kitApi?.getCurrentUser(withKit: self)
        }
        
        func setDeviceToken(_ deviceToken: Data) -> MPKitExecStatus {
            IterableAPI.register(token: deviceToken)
            return MPKitExecStatus(sdkCode: MPKitIterable.kitCode(), returnCode: .success)
        }
        
#if os(iOS)
        @available(iOS 10.0, *)
        func userNotificationCenter(_ center: UNUserNotificationCenter, didReceive response: UNNotificationResponse) -> MPKitExecStatus {
            IterableAppIntegration.userNotificationCenter(center, didReceive: response, withCompletionHandler: {})
            return MPKitExecStatus(sdkCode: MPKitIterable.kitCode(), returnCode: .success)
        }
#endif
        
        func receivedUserNotification(_ userInfo: [AnyHashable: Any]) -> MPKitExecStatus {
            IterableAppIntegration.application(UIApplication.shared, didReceiveRemoteNotification: userInfo, fetchCompletionHandler: { _ in })
            return MPKitExecStatus(sdkCode: MPKitIterable.kitCode(), returnCode: .success)
        }
        
        // MARK: - Accessors
        func advertiserId() -> String? {
            var advertiserId: String?

            if let MPIdentifierManagerClass = NSClassFromString("ASIdentifierManager") as? NSObject.Type {
                let selector = NSSelectorFromString("sharedManager")
                if let adIdentityManager = MPIdentifierManagerClass.perform(selector)?.takeUnretainedValue() as? NSObject {
                    let isAdvertisingTrackingEnabledSelector = NSSelectorFromString("isAdvertisingTrackingEnabled")
                    if let isAdvertisingTrackingEnabled = adIdentityManager.perform(isAdvertisingTrackingEnabledSelector)?.takeUnretainedValue() as? NSNumber, isAdvertisingTrackingEnabled.boolValue {
                        let advertisingIdentifierSelector = NSSelectorFromString("advertisingIdentifier")
                        if let advertisingIdentifier = adIdentityManager.perform(advertisingIdentifierSelector)?.takeUnretainedValue() as? NSUUID {
                            advertiserId = advertisingIdentifier.uuidString
                        }
                    }
                }
            }

            return advertiserId
        }


    }
}
